### PR TITLE
Mirroring infrastructure

### PR DESCRIPTION
**Pull request by akoehn**
This PR should be fairly clean and I hope I documented everything enough. Comments welcome.

### CODE REVIEW

File: Makefile
Comment (mbollmann) [inline Makefile:207]: That backslash at the end seems to be why the build is failing; I guess it's a typo?

File: bin/anthology/utils.py
Comment (mbollmann) [inline bin/anthology/utils.py:230]: Have you checked that using the canonical URL is the correct behaviour for all existing `infer_url` calls? On a glance I'd say yes, but I haven't thoroughly checked

File: bin/anthology/utils.py
Comment (mbollmann) [inline bin/anthology/utils.py:417]: What was the reason again we don't use `infer_url(element.text, template=data.PDF_LOCATION_TEMPLATE)` here? It seems logically equivalent.

File: Makefile
Comment (akoehn) [reply Makefile:207]: Yes, the backslash is wrong. There was an additional line that I deleted before commiting because it was a no-op.

File: bin/anthology/utils.py
Comment (akoehn) [reply bin/anthology/utils.py:230]: Yes, I went through the calls to infer_url.

File: bin/anthology/utils.py
Comment (akoehn) [reply bin/anthology/utils.py:417]: because infer_url creates a canonical URL for pages, not for PDFs.

For example, if you host a mirror, `infer_url("2020.acl-main.1")` should return `https://aclweb.org/anthology/2020.acl-main.1` because that is the canonical URL, but PDFs should link to the mirror and not the primary site.

Comment (mbollmann) [reply bin/anthology/utils.py:417]: But isn't that what the `template=` argument is for?

Comment (akoehn) [reply bin/anthology/utils.py:417]: Possibly.  It is the same with attachments, but I did not write that code and would have to look at it more closely.  Maybe add an issue to refactor that?

**Comment by akoehn** [comment]
updated.  The current code also generated the mirror at http://aclanthology.lst.uni-saarland.de.

File: bin/create_mirror.py
Comment (mjpost) [inline bin/create_mirror.py:143]: Any chance this could be refactored to use `anthology.util.retrieve_url`?

**Review by mjpost — COMMENTED** [review]
Two other issues:

* Can you update the top-level README file with build commands for the canonical site and a mirror?

* I am worried about search indexing. It seems that mirrors should either (a) use a `robots.txt` to prevent crawling or (b) use the site index (or some other mechanism) to point to the canonical variant of all pages.

**Comment by mjpost** [comment]
Related, what about fixing the build issue, perhaps moving to a Git Workflow? We could combine this with publishing each dev branch to `aclanthology.org/dev/{branch}` using the new mirror capability.

Then we could prioritize merging this branch, and merge it into all the failed builds: 
image

File: bin/create_mirror.py
Comment (akoehn) [reply bin/create_mirror.py:143]: Is that comment meant for the downloading? seems to be at an incorrect line.

I used `urlretrieve` because 1) it is built-in in python 2) `retrieve_url` has functionality we don't need (cookies) and does not have functionality we need (https) and has overhead over `urlretrieve`.

**Comment by akoehn** [comment]
>     * Can you update the top-level README file with build commands for the canonical site and a mirror?

Will do later.

>     * I am worried about search indexing. It seems that mirrors should either (a) use a `robots.txt` to prevent crawling or (b) use the site index (or some other mechanism) to point to the canonical variant of all pages.

The meta fields point to the canonical URL but I am not sure whether this works cross-domain.  We can of course ask people to add a robots.txt but I don't know whether that is needed given that people tend to link to the aclweb.org anthology main site.

**Comment by akoehn** [comment]
build issue:  We should move that to the GitHub CI, but IMO that is not that much connected to this PR; the PR previews can easily come later (and we should think about problems such as external people doing PRs and us automatically hosting the contents).

**Comment by mjpost** [comment]
I agree the previews can be seen as separate. I see it as related to Github CI because we currently use that for publishing, and getting publishing for previews working, fits in with the mirroring. But I agree that the previews should not hold up this PR if you don't have time to add it.

However, the broken Travis CI builds are holding everything up, since nothing can be merged even if approved. I was thinking we could fix that by moving to Github. If you don't have time for that, are you able to look into fixing Travis?

**Comment by mjpost** [comment]
> The meta fields point to the canonical URL but I am not sure whether this works cross-domain. We can of course ask people to add a robots.txt but I don't know whether that is needed given that people tend to link to the aclweb.org anthology main site.

They do tend to, but it will still get crawled, and could attenuate Google's signal to canonical URLs. People may accidentally start linking to mirrors, say if the main site is down or they just access the local one. 

I think we should get this right now, before we have ten old mirrors of the Anthology, sapping up link signal.

**Comment by akoehn** [comment]
We can add an additional variable denoting whether we build the official anthology site or a mirror and add a robots.txt if we build a mirror. (we can't check the given prefix because it only denotes where to look for PDFs and such, people might have a mirror that still links to the aclweb.org anthology for PDFs and attachments)

> However, the broken Travis CI builds are holding everything up, since nothing can be merged even if approved. I was thinking we could fix that by moving to Github.

I agree and I think that we should move to the GitHub CI infrastructure ASAP.  However,  1) I don't have the access to modify the CI settings for this repo afaict and 2) we can simply do the same check we currently do on travis, but on GitHub.  The pushing rendered pages out to a webserver can be added later and doing it in two steps is not more work than doing it in one.

Yes, my time is currently unfortunately limited, this PR is my relaxing time :-)

**Comment by mjpost** [comment]
[This](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing) is why our builds are failing: Travis needs money from us. I can't access the acl-org account page so I've escalated to ACL IT.

image

**Comment by mjpost** [comment]
@akoehn do you want to do a dummy push to retrigger the build?

**Comment by akoehn** [comment]
> @akoehn do you want to do a dummy push to retrigger the build?

No, I will push a real change later, but have no time to finish this atm

**Comment by mjpost** [comment]
Oh, even better. No problem, and no hurry.

**Comment by mjpost** [comment]
Just noting this here: we can add demo sites by doing something like this in `.github/workflows/publish.yml`

```yaml
    - name: publish
      env:
        PUBLISH_TARGET: ${{ secrets.PUBLISH_TARGET }}/${GITHUB_REF##*/}
      run: echo Mirroring branch to ${GITHUB_REF##*/}
      run: |
        rsync -aze "ssh -o StrictHostKeyChecking=accept-new" --delete build/anthology/ $PUBLISH_TARGET
```

(although this does that for `master` too)

**Comment by mjpost** [comment]
Hey @akoehn—maybe you're offline for the holidays, but if not and there's anything I can do to help here, please let me know! I'll probably have some time for this this week.

**Comment by akoehn** [comment]
Thanks for asking! I’m not working atm and might add the last few things
(documentation, instructions for updating the mirror via cron-job,
adaptation to the aclweb.org server) when I have some time to relax.  I
don’t think there is that much for you to do.

If you have time, you could maybe look at the missing files I found
working on the mirroring infrastructure (list is in #598)


All the best to all volunteers getting this email :-)

**Comment by mjpost** [comment]
Okay, I think I'm caught up on this.

**Comment by akoehn** [comment]
Short update: I think I have everything in place but I could not test it properly yet; the virtual machine I use for the mirror has only 1GB of RAM, not enough to build the anthology.  Will push my changes soon.

**PLEASE DON'T MERGE THEM YET THEY NEED DISCUSSION**

**Comment by mjpost** [comment]
Awesome, this is exciting! I'm really looking forward to having per-branch site previews.

If it would be helpful, tell me how to build it, and I can do a test locally.

**Comment by akoehn** [comment]
I updated the README with the relevant documentation.  Note that most of the work here is not for per-branch previews as it mirrors all papers etc.

In short: set `ANTHOLOGY_PREFIX` to the URL you want to serve under, `ANTHOLOGYFILES` to where your anthology_files directory is, then run `make` (or `make mirror` if you want to mirror the papers). Let me know if the readme needs more detailed instructions.

**Comment by mjpost** [comment]
Will test tonight.

(I realize this PR doesn't have per-branch mirroring, but it's a dependency)

**Comment by mjpost** [comment]
The `sed -i` command for doing substitution on the `.htaccess` doesn't work on OS X. I'm testing using Perl instead—if that works, I can push up a fix.

**Comment by mjpost** [comment]
I tried to test root-level builds with this command:

```
ANTHOLOGY_PREFIX="https://aclanthology.org" make
```

It built fine, but then when I run `make serve` and load `localhost:8000`, I get this:

image

If I create a dummy symlink `ln -s . anthology` in `build/website`, everything works correctly, including not required `anthology/` in the URL. So I'm not sure what's happening at the top level.

**Comment by akoehn** [comment]
I'm pretty sure that this is a caching problem on your browser's side.  It still assumes that localhost:8000 is a redirect to anthology (from the cache) then tries to load that, then fails. On a fresh browser everything works as expected.

**Comment by mjpost** [comment]
Ah, looks like you're right!

**Comment by mjpost** [comment]
New issue: using a too-deep prefix seems to cause a problem:

```
$ ANTHOLOGY_PREFIX="https://aclanthology.org/previews/mirroring-infrastructure" make
[snip]
INFO     Creating page templates for Hugo...
. "venv/bin/activate" && python3 bin/create_hugo_pages.py --clean
Illegal division by zero at -e line 1, <> line 1.
```

This doesn't happen if I do only one deep, e.g., `ANTHOLOGY_PREFIX="https://aclanthology.org/mirroring-infrastructure" make`

**Review by mjpost — APPROVED** [review]
Looks good! This passes my tests, apart from the remaining issue that specifying a target directory more than one level deep seems to fail.

**Comment by mjpost** [comment]
@akoehn, any reason not to merge this? If you have an idea how to fix the depth > 1 dir issue, we could sneak that in, but if you're short on time, I'd be happy to have this merged. I'd like to get site previews in place soon.

**Comment by akoehn** [comment]
Time to test it. I'm unfortunately seriously lacking time due to kindergartens closing etc.

**Comment by mjpost** [comment]
Found the problem: Makefile uses Perl to do a regexp replacement:

```bash
        @perl -pi -e "s/ANTHOLOGYDIR/$(ANTHOLOGYDIR)/g" build/website/index.html
```

but if `$ANTHOLOGYDIR` contains slashes, this fails. I pushed up a fix. I suppose we should check to ensure that ANTHOLOGYDIR doesn't have `#`s now.

**Comment by mjpost** [comment]
@akoehn I've made a few changes that I'd like to run by you (hopefully you have time, it just requires reading and confirming):

* I got rid of the host and host path as secrets. Having these be secret, together with `rsync --delete`, is a recipe for destruction, particularly since there is no real way to document their expected values. In fact, setting the target path incorrectly is how I wiped out the entire Anthology last year.
* It seems the website is now built into `build/website/anthology`, so I updated the source path for rsync in both the Makefile and in the Github actions
* I created a "preview" Github action that uses the mirroring code to build a website preview for every branch, under https://aclanthology.org/previews/$BRANCHNAME. It was super easy (well, the part that depended on your mirroring code was; getting the Github action syntax right took some work)

Additionally, I propose to add one thing to this branch:

* Update the Github actions to use a Makefile target. This means we only have to maintain syncing code in one place (the Makefile).

**Comment by mjpost** [comment]
I'm done here. The previews are working! See here for example: https://aclanthology.org/previews/mirroring-infrastructure/.

I tried to get this working so that the preview could be pushed out as a separate workflow after `check-build`, but got tired of fiddling with it. One question I couldn't figure out is if the machine-state is preserved across workflow tasks that have been linked. This would be easier to test with some dummy workflows that finish in 2s.

I've also add an action that will remove the preview when the branch is removed.

It'd be nice to add one more workflow that adds a comment to each PR, providing a link to the preview on aclanthology.org!

**Comment by akoehn** [comment]
@mjpost looks good! If you want, I can have an in-depth look at it tomorrow; otherwise you can just merge it.

**Comment by mjpost** [comment]
I decided to merge, since this is an old PR, and my changes were just minor on top of yours. We can always take it up in future PRs.

**Pull request merged**
